pkg/agent/tcpconn/queue.go

```diff
@@ -0,0 +1,172 @@
+package tcpconn
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"math/rand"
+
+	"github.com/florianl/go-nfqueue"
+	"github.com/grafana/xk6-disruptor/pkg/iptables"
+	"github.com/grafana/xk6-disruptor/pkg/runtime"
+)
+
+// NFQConfig contains netfilter queue IDs that iptables rules and the packet handler use to communicate.
+type NFQConfig struct {
+	// QueueID is an arbitrary integer used to identify a queue where a handler listens and a disruptor redirects target
+	// packets.
+	QueueID uint16
+	// RejectMark is an arbitrary integer which the handler uses to mark packets that need to be dropped.
+	RejectMark uint32
+}
+
+// RandomNFQConfig returns a NFQConfig with two random integers to be used as queue IDs and reject mark.
+// To ensure the numbers are not zero they are ORed with 0b1, as adding 1 can actually result in the number overflowing
+// and becoming zero.
+func RandomNFQConfig() NFQConfig {
+	return NFQConfig{
+		QueueID:    uint16(rand.Int31()) | 0b1,
+		RejectMark: uint32(rand.Int31()) | 0b1,
+	}
+}
+
+// Queue is the interface implemented by objects that can write packets to a channel.
+type Queue interface {
+	// Start starts listening to packets, blocking until an error occurs or the context is cancelled.
+	Start(ctx context.Context, packets chan<- Packet) error
+}
+
+// Packet is a packet received by a Queue. It can describe itself and be accepted or rejected.
+type Packet interface {
+	// Bytes returns the raw bytes that make the packet.
+	Bytes() []byte
+	// Accept marks the packet to be accepted.
+	Accept()
+	// Reject marks the packet to be rejected.
+	Reject()
+}
+
+// FakeQueue is a fake implementation of queue. It stalls until the context is cancelled, never sending any packet.
+type FakeQueue struct{}
+
+// Start implements the Queue interface.
+func (f FakeQueue) Start(ctx context.Context, _ chan<- Packet) error {
+	<-ctx.Done()
+	return ctx.Err()
+}
+
+// NFQueue implements the Queue interface using netfilter's nfqueue mechanism, reading packets sent to userspace by
```

Comment:
I'm not opposed to change this, but long text on why I initially chose to do it this way:

> There's something strange in the way responsibilities are presently split between the disruptor, the dropper, and the queue.

My main goal with designing these three in the way they are were to to abstract out the details of how the queue works (e.g. the nftables-related stuff). The responsibilities as of now would be:
- The dropper reconstructs a connection from a packet and decides what to do with it
- The queue produces (but does not consume) *actionable* packets
- The disruptor consumes packets and uses the dropper to action them

This way, the disruptor does not need to care about where the packets come from or how they are actioned, at the cost of the what I admit to be a bit convoluted flow (specially with the NFPacket).

That being said, I do not oppose to simplifying the flow by consuming the packets in the Queue, but the cost of this would be that the queue would grow in responsibilities. LMK what your thoughts on this tradeoff are.

Reply:
I understand the rationale, but I think we are introducing too much complexity for a benefit that is not really materializing in this context. I don't see any benefit of having the disruptor as an intermediary, as the decision is actually made by the dropper. 

I would prefer to simplify the code.

Reply:
Makes sense, I'll move the packet consumption logic to the queue and see how that looks.

Reply:
I've refactored this in the latest commits of the PR. When I moved the packet processing logic to the queue, I realized the disruptor was not doing anything anymore, just calling Queue.Start. For this reason I just went away and merged the queue into the disruptor.

Along the way I also slightly split the disruption into two (Disruption and Filter), which I think it makes the separation more clear: The Disruptor no longer needs the Disruption object, only the Dropper that is built from it. What it cares about is the Filter, as it needs it to set up the relevant rules.

Looking forward to your feedback for these changes :)

Reply:
I like these changes. The code now is more compact and easier to follow, and we are saving copying the message to a channel, which is not a minor improvement.